core/trino-main/src/main/java/io/trino/execution/SqlQueryExecution.java

```diff
@@ -129,6 +134,9 @@
     private final CostCalculator costCalculator;
     private final DynamicFilterService dynamicFilterService;
 
+    @Nullable
+    private Thread planningThread;
```

Comment:
It is used in async lambdas and cannot be a variable. So I could either decorate it in some `AtomicReference` type of object of make it a field.

Reply:
It is being set to `null` later on

Reply:
It's not clear to me why we need that. The only reason would be to prevent the wrong thread from being interrupted, but that can't really happen. If the query transitions to FAILED during planning, it *must* still be executing the `start()` method, since all the planning methods are synchronous. It is only then that the thread is interrupted, which means it will be the right thread.

I can't think of a combination of actions that would result in `start()` completing and the query not already be in a state other than PLANNING, which means the interruption wouldn't happen beyond that point.

Reply:
if we remove `planningThread`, we also remove synchronization around setting/unsetting it. Wouldn't that introduce race conditions? Would we be able to still reason that `start` didnt compete yet?

Reply:
The synchronization doesn't prevent any race conditions -- it's only for JMM semantics regarding updating/reading that field from multiple threads.

In any case, I think it's worth adding a comment along the lines of what I described above. Reasoning about concurrency is hard, and we don't have the right primitives here to make it easier.

Reply:
This was implemented to prevent the interruption in an unlikely case of the thread being already reused for another query. This can be simulated by placing long sleep just before the `interrupt()` method. That way we can interrupt the next query executed in this thread.

Reply:
State change listener can execute any time in the future (even when `planningThread` started serving another query)

Reply:
Fair enough. The use of `directExecutor()` in the addListener call threw me off.

However, if can happen *only* if the planning fails from the call to `fail()` in the catch at the bottom of this method.

Still, I would write this code as follows to keep the state and logic self-contained and not have to reason about a class-level field, it's lifetime beyond this method (or other methods, etc). Ideally, we should be able to remove listeners, but there's no facility to do so:

```java
final AtomicReference<Thread> planningThread = new AtomicReference<>(currentThread());
stateMachine.getStateChange(PLANNING).addListener(() -> {
    if (stateMachine.getQueryState() == FAILED) {
        Thread thread = planningThread.get();
        if (thread != null) {
            thread.interrupt();
        }
    }
}, directExecutor());

try {
    PlanRoot plan = planQuery();
    // DynamicFilterService needs plan for query to be registered.
    // Query should be registered before dynamic filter suppliers are requested in distribution planning.
    registerDynamicFilteringQuery(plan);
    planDistribution(plan);
}
finally {
    planningThread.set(null);

    // clear the interrupted flag in case there was a race condition where
    // the planning thread was interrupted right after planning completes above
    Thread.interrupted();
}
```

Reply:
> However, if can happen only if the planning fails from the call to fail() in the catch at the bottom of this method.

Why only then? Query might be interrupted via timeout and `thread.interrupt();` might be delivered in arbitrary time after that.

Other than that keeping logic self-contained within method makes sense

Reply:
> Why only then? Query might be interrupted via timeout and thread.interrupt(); might be delivered in arbitrary time after that.

Ah yes, scratch that. I was thinking of the case where the interrupt causes those methods to throw.